dialogue-core/src/main/java/com/palantir/dialogue/core/RetryingChannel.java

```diff
@@ -46,61 +44,59 @@
-        private final IntFunction<ListenableFuture<Response>> runnable;
-        private final SettableFuture<Response> delegate;
+    private static final class RetryingCallback {
+        private final Channel channel;
+        private final Endpoint endpoint;
+        private final Request request;
+        private final int maxRetries;
+        private int failures = 0;
 
-        private RetryingCallback(IntFunction<ListenableFuture<Response>> runnable, SettableFuture<Response> delegate) {
-            this.runnable = runnable;
-            this.delegate = delegate;
+        private RetryingCallback(Channel channel, Endpoint endpoint, Request request, int maxRetries) {
+            this.channel = channel;
+            this.endpoint = endpoint;
+            this.request = request;
+            this.maxRetries = maxRetries;
         }
 
-        @Override
-        public void onSuccess(Response response) {
+        ListenableFuture<Response> execute() {
+            return wrap(channel.execute(endpoint, request));
+        }
+
+        ListenableFuture<Response> success(Response response) {
             // this condition should really match the BlacklistingChannel so that we don't hit the same host twice in
             // a row
             if (response.code() == UNAVAILABLE_503 || response.code() == TOO_MANY_REQUESTS_429) {
                 response.close();
-                retryOrFail(Optional.empty());
-                return;
+                return retryOrFail(Optional.empty());
             }
 
             // TODO(dfox): if people are using 308, we probably need to support it too
 
-            boolean setSuccessfully = delegate.set(response);
-            if (!setSuccessfully) {
-                response.close();
-            }
+            return Futures.immediateFuture(response);
         }
 
-        @Override
-        public void onFailure(Throwable throwable) {
-            retryOrFail(Optional.of(throwable));
+        ListenableFuture<Response> failure(Throwable throwable) {
+            return retryOrFail(Optional.of(throwable));
         }
 
-        private void retryOrFail(Optional<Throwable> throwable) {
-            int attempt = failures.incrementAndGet();
+        private ListenableFuture<Response> retryOrFail(Optional<Throwable> throwable) {
+            int attempt = ++failures;
             if (attempt <= maxRetries) {
-                DialogueFutures.addDirectCallback(runnable.apply(attempt), this);
-            } else {
-                if (throwable.isPresent()) {
-                    delegate.setException(throwable.get());
-                } else {
-                    delegate.setException(new SafeRuntimeException("Retries exhausted"));
-                }
+                return execute();
             }
+            return Futures.immediateFailedFuture(
+                    throwable.orElseGet(() -> new SafeRuntimeException("Retries exhausted")));
```

Comment:
It might be nice to know what caused us to retry

Reply:
good call, might as well add some helpful info logging prior to each retry while I'm here.